Add functionality to Podcast/Episode Textviews in CoverFragment

ueen opened the pull request:
discussion https://github.com/AntennaPod/AntennaPod/pull/5067
So this is my clean, tested PR with this behaviour
- PodcastTextview onClick: Opens Feed
- EpisodeTextview onClick: Opens Share Dialog
- Both onLongClick: Copies Title into Clipboard

Tell me what you think, hope the tests all check out.

ByteHamster commented:
> Alright, sorry for bothering so much, i have a working build but the issue is this: `textIsSelectable=true` and `background=?android:selectableItemBackground` lead to strange behaviour: The Textview after clicked is "selected" (halfwhite selectableItemBackground) and reacts to a second click with the desired action

Pretty weird behavior but I can confirm that it happens. I have never seen something like it.

> EpisodeTextview onClick: Opens Share Dialog

To be honest, this feels pretty strange to me. The title of an episode has nothing to do with sharing. The text could just as well add it to the queue, mark it as favorite or add it to the queue. All of those options are not really related to the title text. So I would remove that. /CC @keunes

> PodcastTextview onClick: Opens Feed

After trying it a few times... do we really need two different ways of navigating to the podcast screen that are displayed on the audio player? Google will probably complain about the touchable area being too small but I wouldn't make it larger just because of that click action.

ueen commented:
IMHO it does makes sense that clicking on the title is opening the share dialog (because you are selecting the episode, proabably to share it, why else would you click it?). Personally i never hat the urge to select the text of the Podcast/Epsiode title and tbh was quite suprised that this is the case and actaully thought it was a bug at first.

As i said above an alternative could be to open another dialog with severeal options when clicking on the episode, as there are only two sensible (open website and share) it wouldnt make to much sense, and why would you want to open the website if not to share it?

I think copy onLongclick is also better because it is just von longclick instead now: longclick, select all text, press copy.

This is convenience behaviour for sure, but i think they would really benefit the day to day experience, 
the benefit of naviagting to the feed via podcasttextview rather than menu is that it is
- one click
- closer to the bottom (one hand friendly)

Thanks for your constructive feedback :)

ueen commented:
Ok i guess the sharing functionaity is not so important to me, but i do need an **easier way to get to the feed**, maybe **clicking the cover**? That would make sense no? Maybe this could be customizable also, i get your point about not having to many options, but i just discovered FairEmail and its actually kind of awsome to tweak it just how its most useful to you :)

keunes commented:
> i just discovered FairEmail

I use it also because I don't know of another FOSS app, but it's configuration I find a nightmare.

> clicking the cover?

I think that's already used for play/pause. I don't use it myself, but I think it doesn't merit a setting. But isn't the podcast view already fairly accessible (like, in two taps)?

ueen commented:
Then I'll recommend K9 mail to you, if you don't like the old UI you can enable the beta with the new decent looking UI, I know the configurations of FairEmail are a lot, I don't want to turn AP that way, but I actually do love the way I'm able to use FairEmail after tweaking all the settings.

So yea often I listen to a podcast and afterwards want so select another episode, and with one hand it's rather difficult to reach the top menu and I think the cover for playpause is rather redundant, because the playpause button is right there! Also it makes sense, as the cover is the one of the whole podcast->all episodes, this could be a good solution to the this issue because clicking on the titles is indeed difficult due to the small clicking area.

ByteHamster commented:
> I think the cover for playpause is rather redundant

It was requested by users who use AntennaPod in a car and need to play/pause without looking at the screen. 

> as the cover is the one of the whole podcast->all episodes

Covers are episode specific. The screen also shows chapter images if the podcast creator supports that.

---

I could live with the following:

- Click or long click on the episode title: Copy the title to the clipboard (and show a snackbar telling that it copied)
- Click on the podcast title: Go to the podcast screen
- Long click on the podcast title: Copy the title to the clipboard (and show a snackbar telling that it copied)
- Remove the item in the toolbar on top

Adding menus to a text label that, by design, isn't even supposed to do anything is an overkill in my opinion. @keunes @ueen how do you feel about above behavior?

ueen commented:
Adding menu in the sense of a Dialog that displays the options, I think that's common behavior no? 

I can live with the title->feed and long click copy as well, maybe I can refactor the code to do this cleanly, maybe I could add a small padding of 2dp or so to increase the touch area a little bit without curbing aesthetics?

ByteHamster commented:
> I think that's common behavior no?

I mainly meant that generic text labels somewhere on the screen shouldn't have too much functionality - because it's not expected.

> maybe I could add a small padding of 2dp or so to increase the touch area a little bit without curbing aesthetics?

Not sure, could work. We need to be careful about not to make the screen too full. Especially when discussing about adding even more text to the screen (#5075).

ueen commented:
Ok agreed, I'll test.

ueen commented:
Alright, i removed the onClick from the episode title, now it should work as you described above :)

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/CoverFragment.java:
```diff
@@ -226,4 +240,13 @@ void onPlayPause() {
+        if (clipboardManager != null) {
+            clipboardManager.setPrimaryClip(ClipData.newPlainText("AntennaPod", text));
+        }
+        Toast.makeText(requireContext(), getResources().getString(R.string.copied_to_clipboard), Toast.LENGTH_SHORT).show();
```
Please use a Snackbar instead to be consistent with other parts of the app. We have a method called `((MainActivity) getActivity()).showSnackbarAbovePlayer` for that.

ByteHamster commented on app/src/main/java/de/danoeh/antennapod/fragment/CoverFragment.java:
```diff
@@ -13,23 +16,30 @@
+import de.danoeh.antennapod.activity.MainActivity;
 import de.danoeh.antennapod.core.event.PlaybackPositionEvent;
+import de.danoeh.antennapod.core.feed.FeedItem;
+import de.danoeh.antennapod.core.feed.FeedMedia;
```
The package of those imports was changed in the meantime (I know, that type of change is pretty annoying). This causes the build to fail, currently. Could you please fix this, so the CI checks get green?

ueen commented:
Ok, wanna merge, or should i merge into https://github.com/AntennaPod/AntennaPod/pull/5075? :)

ByteHamster commented:
I would say let's wait with this one until #5075 is merged. Then each PR focuses on one single feature.

ueen commented:
Alright, the other PR is merged :)
So i added a way to see the whole title if its to long (scrolling as of now is really hard) so i added this nice marquee efffect:

https://user-images.githubusercontent.com/5067479/118364963-05651100-b59b-11eb-8d0e-9e59a84848e5.mp4

ByteHamster commented:
> scrolling as of now is really hard

Yeah, that's not even an actual feature but it somehow works :D

> marquee efffect on click

I wonder if we should automatically do the effect, so you don't have to click it. I don't think having to click it is very discoverable. Always scrolling could also be annoying, though. Not sure.

> Alright, the other PR is merged :)

Could you please resolve the merge conflicts?

ueen commented:
> I don't think having to click it is very discoverable. Always scrolling could also be annoying, though.

Yea thought about that as well, maybe we can keep it onClick for now? We can always change it later on, for now this just adds the possiblility to see the full title :)

ByteHamster commented:
Thanks! Will be released in version 2.3.0
